src/yetibot/commands/google.clj

```diff
@@ -0,0 +1,245 @@
+       :args (convert-keys-into-google-keywords (:options output))})))
+
+(defn common-messaging-interface
+  [status message & {:keys [results]
+                     :or {results nil}}]
+  {:status status
+   :message message
+   :results results})
+
+(defn search-function-input-validation
+  [query]
+  (let [proc-query (options-processor query)
+        messagef common-messaging-interface]
+    (cond
+      (:errors proc-query) (messagef :failure (join "\n" (:errors proc-query)))
+      (empty? (:query proc-query))
+                           (messagef :failure (:empty-query messages))
+      :else
+      (messagef :success nil :results proc-query))))
+
+(defn common-search-function
+  [query & {:keys [sfunction type]
+   :or {sfunction api/search type :normal}}]
+  (let [validation-results (search-function-input-validation query)]
+    (condp = (:status validation-results)
+      :failure validation-results
+      (let [messagef common-messaging-interface
+            proc-query (:results validation-results)
+            args (:args proc-query)
+            results (sfunction (:query proc-query)
+                               :args (merge @options args))
+            count (get-in results [:searchInformation
+                                   :totalResults])]
+        (condp = count
+          nil (messagef :failure (:google-died messages))
+          "0" (if (= type :image)
+                (messagef :failure (:image messages))
+                (messagef :failure (:search messages)))
+          (messagef :success nil :results results))))))
+
+(defn load-options-from-file-into-atom
+  []
+  (let [options  (api/populate-options-from-config)]
+    (if (empty? options)
+      (info (:no-options-in-config-file messages))
+      (doall (map (fn [[k v]]
+                    (if-let [result (set-option-value k v)]
+                      (info result)
+                      (info k ":" (:option-not-found messages))))
+                  options)))))
+
+(load-options-from-file-into-atom)
+(defonce copy-of-valid-file-options
+  @options)
+
+(defn search
+  "google search <query> # plain google search"
+  [{[_ query] :match}]
+  (let [results (common-search-function query)]
+    (if (= :failure (:status results))
+      (:message results)
+      (api/format-results (get-in results [:results :items])))))
+
+(defn image-search
+  "google image <query> # image search"
+  [{[_ query] :match}]
+  (let [results (common-search-function query
+                                        :sfunction api/image-search
+                                        :type :image)]
+    (if (= :failure (:status results))
+      (:message results)
+      (api/format-results (get-in results [:results :items]) :order :image))))
+
+(defn option-info
+  "google option-info <query> #display info for option"
+  [{[_ query] :match}]
+  (if-let [info (get-info (keyword query))]
+    info
+    (:option-not-found messages)))
+
+(defn show-custom-keywords-list
```

Comment:
fn name should match the command, e.g. `show-options` or `show-options-cmd`. Also, we could just shorten this to `google options` which means the fn would be named `options` or `options-cmd`, and it could return a vector of options so that it would be easy to use with `xargs`. That would allow a user to quickly see all option docs using:


```
!google options | xargs google option-info
```

See [room](https://github.com/devth/yetibot.core/blob/master/src/yetibot/core/commands/room.clj) and [category](https://github.com/devth/yetibot.core/blob/master/src/yetibot/core/commands/category.clj) commands for similar examples / inspiration.